genmake2 for hybrid adjoint modeling

IvanaEscobar opened the pull request:
## What changes does this PR introduce?
A way to compile `mitgcmuv_ad` with codes that contain free-format FORTRAN that use modules. 
This may address a few points from issue [855](https://github.com/MITgcm/MITgcm/issues/855). Specifically these

> * [ ]  changes to be able to use F90 "module" in src code sent to TAF.
...
> * [ ]  more logical and consistent ordering inside targets: ".F77", ".F90", ".c", ".h" ...  Currently the F90 is sometimes pushed after the C-src code

## What is the current behaviour? 
We cannot compile a hybrid adjoint model with the current `MITgcm/tool` scripts


## What is the new behaviour 
We can use `make adall` to create a TAF generated AD model and compile `mitgcmuv_ad`. I swapped order to prioritize F90 code first so that we do not need to worry of F77 module dependencies as well.  

**NOTE:** Only works with `genmake2` flags `-nocat4ad | -dog4ad | -ncad | -dad`, which are for sending multiple files to TAF instead of a single concatenated file such as `ad_input_code.F`

Changes also handle: 
- Fix the missing dependence in `Makefile` from TAF generated F90 code

## Does this PR introduce a breaking change? 
Unaware if these points break code.
- Manual addition of `-e` in the TAF command. Previously, I was using `USE_EXTENDED_SRC='t'` in my `genmake_local`. This was placing `-e` in the wrong spot
- `genmake2` requires bash commands `tr` and `tsort` which were available on all the machines I have access to. 
- `genmake2` has in-line `sed` edits to several `$TMP.*` files, e.g. `sed -i ...`. This feature was avoided in `genmake2` and I wasn't sure if there was a reason to do so. 
- `FLOWFILES` were left to be passed to TAF as `-fixed`. It doesn't seem to matter where they go. 

## Other information:
This routine was tested on MacOS M2 and several Linux systems. I used `checkpoint68s` and TAF version 6.4.4 . This needs to be verified with a simpler example that contains F90 Modules.

These changes use `f90mkdepend` from PR [858](https://github.com/MITgcm/MITgcm/pull/858)

## Suggested addition to `tag-index`
`genmake2` compiles TAF AD with module enabled hybrid codes.

IvanaEscobar commented:
Doesn't work with 
```
make -j N adall
```

IvanaEscobar commented:
Checked on MacOS, and had to remove inline `sed` editing. 

Also rearranged forward target order to match changes made to AD target order. 

In MacOS, I was running into a `rm` error that required a recursive removal of `genmake_tc*` directories. 

Did not include the `f90mkdepend` changes applied in PR [858](https://github.com/MITgcm/MITgcm/pull/858)

mjlosch commented:
@IvanaEscobar do you have a simple example, where I can test this? 

`verification_other/atm_gray` has no AD experiment, and even the forward code does not compile with this PR.

IvanaEscobar commented on tools/genmake2:
```diff
@@ -881,7 +881,7 @@ EOF
 
     #  cleanup the testing files
-    rm -f genmake_tcomp.* genmake_test.*
+    rm -rf genmake_tcomp.* genmake_test.*
```
I had to add it when compiling in MacOS. Otherwise, the directories created weren't being removed.

IvanaEscobar replied:
This has stopped being the case for forward code compilation with and without `-devel` for my case, using Darwin 24.0. 

Checking with the adjoint TAF

IvanaEscobar commented on tools/genmake2:
```diff
@@ -2774,15 +2774,42 @@ fi
     for j in $list_files ; do
-	grep '\.f\>'   $i/$j >> $TMP.adSrcFiles
-	grep '\.f90\>' $i/$j >> $TMP.adF90Files
+	    grep '\.f\>'   $i/$j >> $TMP.adSrcFiles
+        # Retain file paths for topo sorting later
+	    grep '\.F90\>' $i/$j | while IFS= read -r line; do
+            echo "$i/$line" >> $TMP.adF90Files
+        done
```
At this stage in the code, there are no linked files in your `build` directory. I retain the use of `.F90` in my list because the routine later searches for files with that file ending in the package path. So it's looking for `myfile.F90` in `MITgcm/pkg/mypackage/` instead of in the `build` directory. 

I thought this would be easier to do. An alternative would be change the file endings as you suggested. Is that the preferred way forward?

IvanaEscobar replied:
You answered my question already in the comment below. 

Just have to change the file endings in the top sort segment as back from `F90` to `f90` as well

IvanaEscobar commented on tools/genmake2:
```diff
@@ -2774,15 +2774,42 @@ fi
+    mod_used=$(grep -i "^[^!]*\buse\b" $f90file | awk '{print $2}' | \
+        tr -d ',')
+    
+    for mod in $mod_used; do
+        # Find the file that defines this module
+        mod_file=$(grep -li "^module $mod\b" $(grep -v $f90file "$TMP.adF90Files"))
+        if [ -n "$mod_file" ]; then
+            echo $mod_file $f90file >> "$TMP.adF90Graph"
+        fi
+    done
```
I see. The topological ordering only matters for files that use modules. The rest can be appended afterwards. 

This is missing logic to include _non-module_ F90 source codes.

IvanaEscobar replied:
I'm weary of adding too many new temporary files, which was my first idea on how to handle a _non-module_ list to be appended _after_ `$TMP.F90Graph` is written in the correct order to `$TMP.adF90Files`. 

Let me know if you think it's a non-issue @mjlosch

mjlosch replied:
I agree that it is better to append the non-module f90-files after the `$TMP.F90Graph` is written in the correct order to `$TMP.adF90Files`, rather than creating a new temporary file.

IvanaEscobar commented on tools/genmake2:
```diff
@@ -3430,7 +3491,7 @@ done
 rm -f $TMP.adSrcFiles
 printf "AD_F90FILES = " >> $MAKEFILE
-AD_F90FILES=`cat $TMP.adF90Files | sed -e "s/\.f90/.$FS90/g"`
+AD_F90FILES=`cat $TMP.adF90Files | sed -e "s/\.F90/.$FS90/g"`
```
got it. I'll move the update to the earlier comment

IvanaEscobar commented on tools/genmake2:
```diff
@@ -3218,7 +3279,7 @@ cat $TMP.F90srclist      >> $MAKEFILE
 cat $TMP.ad_flow_files   >> $MAKEFILE
 
-rm -f $TMP.F77srclist $TMP.nonADF77srclist $TMP.nonADF90srclist $TMP.csrclist $TMP.F90srclist $TMP.hsrclist $TMP.ad_flow_files
+#rm -f $TMP.F77srclist $TMP.nonADF77srclist $TMP.nonADF90srclist $TMP.csrclist $TMP.F90srclist $TMP.hsrclist $TMP.ad_flow_files
```
cleaning that up

IvanaEscobar commented on tools/genmake2:
```diff
@@ -3038,6 +3065,9 @@ for d in $alldirs ; do
 		    F90)
 			echo    " \\" >> $TMP.F90srclist
 			printf " $sf" >> $TMP.F90srclist
+            # Store file path and name for topo ordering later
+            echo "$d/$sf" >> $TMP.F90Files
+
 			if ( test "x$OPENAD" = x && test "x$TAPENADE" = x ) ; then
```
Why does this need to move into the next loop? `$TMP.F90Files` is supposed to be the same thing as `$TMP.F90srclist` just with file paths prepended.

mjlosch replied:
The way I read this is that lines 3065 and 3066 are on the same loop/if level (in the F90 case, 3 tabs from the beginning of the line in my editor) as the `printf " $sf" >> $TMP.F90srclist` statement in the previous line 3064 and the `if ( test "x$OPENAD"` in line 3068 (correct me if I am wrong), so that this comment was just about fixing some indentation.

IvanaEscobar replied:
I misread the suggested change! You're right.

IvanaEscobar commented:
Updated `genmake2` based on some of @mjlosch 's comments. Left some of the comments unresolved for now.

mjlosch commented:
@IvanaEscobar thanks for making the changes. As far as I can see, these two issues remain
1. `genmake2` expects upper case `.F90` suffixes in the `*_ad_diff.list` files and will omit lower case files (I think because l2787 reads ` grep '\.F90\>' $i/$j | while IFS= read -r line; do ...`). After merging #863, there is one example in `hs94.1x64x5` in the current master and this verification experiment will fail, because the module-less file `cost_test_local.f90` in `code_ad_diff.list` will not be recognised and a TAF-error follows: 
```
TAF ERROR    Code of subroutine  cost_test_local not seen and no flow information set,
TAF ERROR    assuming all arguments are input and output
```

2. The logic to include non-module F90 source codes is missing, which a second reason, why the new `hs94.1x64x5` will not compile.

We can wait until #858 and #865 are merged before we continue with this, unless you want to merge `upstream/master` (`git pull upstream master`) into this branch now. Then it would be obvious what I mean.

mjlosch commented on tools/genmake2:
```diff
@@ -3512,62 +3570,58 @@ ad_taf_output.$FS: ad_input_code.$FS
 	@-rm -f ad_input_code_ad.$FS ad_inpF90_code_ad.$FS90 ; echo ''
-	\$(TAF) \$(AD_TAF_FLAGS) \$(TAF_EXTRA) -fixed ad_input_code.$FS -free ad_inpF90_code.$FS90
-	ls -l ad_input_code_ad.$FS ad_inpF90_code_ad.$FS90
+	\$(TAF) \$(AD_TAF_FLAGS) \$(TAF_EXTRA) -free -N 150 ad_inpF90_code.$FS90 -fixed -e ad_input_code.$FS
```
another comment: the `-N 150` appears to be experiment-specific. As far as I know the default line length is 138 and I do not know what the consequences are of having `-N 150` as the default (it's probably harmless, isn't it?). 
I would opt for putting this extra flag into a `genmake_local` (`TAF_EXTRA="-N 150"`)

mjlosch replied:
I was wrong about this, it is the `-e` that is the problem. Again, this should probably be part of a `genmake_local`  (`USE_EXTENDED_SRC`) that is then somehow translated into a `-e`, we already do something like this for `verification/bottom_ctrl_5x5`

IvanaEscobar replied:
I didn't know hoe to best manage `-N 150`. It is specific to an F90 header file hack I use, which isn't standard for MITgcm at this point. 

There can be an additional `TAF_EXTRA_FREE` flag to manage flags that apply only to the free-formatted source code passed to TAF. Does that seem like the best path forward?

IvanaEscobar commented:
> 
> We can wait until #858 and #865 are merged before we continue with this, unless you want to merge `upstream/master` (`git pull upstream master`) into this branch now. Then it would be obvious what I mean.

Waiting for the updated `f90mkdepend` and `xmakedepend` will be best. I see what you mean. Two lists need to exist: one for all non-module F90 source files and another for module-enabled F90 source files.

IvanaEscobar commented:
Merged most F90 related updates. Currently, waiting on #865

IvanaEscobar commented:
@mjlosch  I made some changes I thought were already here, but were in a different repo! 

The next changes have to do with setting `-free` and `-fixed` flags separately. Currently, they are hardcoded to work with my set-up within `genmake2`. Should this be accomplished through the adjoint opt files?

mjlosch commented:
> The next changes have to do with setting `-free` and `-fixed` flags separately. Currently, they are hardcoded to work with my set-up within `genmake2`. Should this be accomplished through the adjoint opt files?

@IvanaEscobar For your experiment you appear to need the TAF-flag sequence ` -free -N 150 $(AD_F90FILES) -fixed -e $(FLOWFILES) $(AD_FILES)`, which is currently hard coded. The extended source lines for fixed format `-e` appears to be necessary, too, for TAF to complete the command.

I managed to use `-N 150 -free $(AD_F90FILES) -fixed -e $(FLOWFILES) $(AD_FILES)` successfully, with `-N 150` before `-free` so that it could be possible that we can put the `-N 150` into the TAF_EXTRA variable in `genmake_local`, right? For some reason the `-e` was still possible.

It's probably best to have two variables in `adjoint_defaults`, that default to `TAF_FREE=-free` and `TAF_FIXED=-fixed` which can be set in `genmake_local` to e.g. `TAF_FREE=-free -N 150` and `TAF_FIXED=-fixed -e` (or `TAF_FREE=-N 150` and `TAF_FIXED=-e` which are then prepended by the `-free` and `-fixed` flags). That should be relatively straightforward to implement. Would you like to try that?

> 
> Another task is to consider non-module using F90 codes.

I can try to address this.

mjlosch commented:
This should now work for all experiments that use `USE_EXTENDED_SRC='t'` in `genmake_local` (currently only `verification/bottom_ctrl_5x5`). To be fixed (see previous posts): make it possible to specify `-free` with `-N 150` and `-fixed` with `-e` for individual experiments via `genmake_local`.

jm-c commented:
To allow specific TAF-Option for fixed and free format files, I think we agree on the solution of adding 2 new variables to be set in the adjoint-optfile. As we discussed earlier today, these could be named "TAF_FIXED_FLAGS" and "TAF_FREE_FLAGS".
Alternatively, we could name them "TAF_F77_FLAGS" and "TAF_F90_FLAGS" as "F77" and "F90" are already used in `genmake2` and option-files to distinguish between fixed and free format.

jm-c commented:
And regarding the changes here from "rm -f" to "rm -rf": if the current "rm -f" does not break the generation of a working `Makefile` (just leaving few dir behind), I would be safer to keep it this way (may be add these few dir to the "make CLEAN" target ?). But if it's necessary to get the `Makefile` then it's something that needs to be fixed.

mjlosch commented:
In commit [01d58eb](https://github.com/MITgcm/MITgcm/pull/862/commits/01d58ebf5947169d012386d26316728f0a5295a0) I went a little further and allowed these flags to be set with options in the `genmake2` command and via `genmake_local`. That's not what we discussed, but it's easy to remove.

IvanaEscobar commented on tools/genmake2:
```diff
@@ -2774,15 +2787,53 @@ fi
+    if test -z "$inList" ; then
+	echo $f >>  $TMP.adF90Files
+    fi
+done < $TMP.adF90noGraph
```
This is great! I hadn't noticed this update until recently. 

This covers `F90noGraph` files assuming that TAF *must* see all F90 files. 

I haven't attempted the case where there is a `F90noGraph` file, that also doesn't need to be TAF'ed. I am unsure if you can simply remove this file from your `_ad_diff.list` and have TAF still develop the ADJ/TLM models (as it does with all of MITgcm's F77 non-AD src codes). 

If an F90 file hold a module *or* if it contains a `USE` statement, then TAF must see the source code. Otherwise, I think the original MITgcm technique of hiding the code can be applied. Does that make sense @mjlosch ?

IvanaEscobar replied:
Should work.

IvanaEscobar commented:
This works for my set-up, and generates `mitgcmuv_ad` as expected. 

I updated my `genmake_local` file to account for the new `TAF_F90/77_FLAGS`.

I also added `NON_AD_F90_SRC_FILES` back into the `LINK`ing command.

IvanaEscobar commented:
> And regarding the changes here from "rm -f" to "rm -rf": if the current "rm -f" does not break the generation of a working `Makefile` (just leaving few dir behind), it would be safer to keep it this way (may be add these few dir to the "make CLEAN" target ?). But if it's necessary to get the `Makefile` then it's something that needs to be fixed.

@jm-c this change isn't required anymore. I've removed them

IvanaEscobar commented:
Thanks @mjlosch , not sure what possessed my text editor

IvanaEscobar commented:
These modifications work for my local set-up. With these modifications the following are addressed:

- Correct files listed in `AD_F90FILES` and `NON_AD_F90_SRC_FILES`
- Ability to use `adall`, `adtaff` and `ftlall`, `ftltaf` with no manual intervention to generate AD and FTL executables
- Apply TAF flags on `-free` independently from `-fixed` parts of the command, handled via `genmake_local` _or_ `genmake2` flags

How does this look for y'all? @jm-c @mjlosch

mjlosch commented:
I think that this looks good now.

mjlosch commented on tools/genmake2:
```diff
@@ -3376,7 +3472,7 @@ CPPCMD = cat \$< | ${CPP} \$(DEFINES) \$(INCLUDES) | ${S64}
 .$FS90.o:
 	\$(F90C) \$(F90FLAGS) \$(F90OPTIM) -c \$<
 .f$FS90.o:
-	cp \$< \$(basename  \$<).f90
+	cp \$< \$(basename  \$<).$FS90
 	\$(F90C) \$(F90FLAGS) \$(F90OPTIM) -c \$(F90FIXEDFORMAT) \$(basename  \$<).f90
```
I am not sure if this is correct because the F90 command still operates on `\$(basename \$<).f90`. shouldn't l3380 be:
```
\$(F90C) \$(F90FLAGS) \$(F90OPTIM) -c \$(F90FIXEDFORMAT) \$(basename \$<).$FS90
```
? When are the `.ff90` files used? Tapenade? OpenAD?

IvanaEscobar replied:
Yes, that would be consistent

mjlosch reviewed (commented):
LGTM.

There are many cases and it's hard to excluded that we haven't missed anything.
Any idea how to address the failing `make -j` problem?

IvanaEscobar commented:
> There are many cases and it's hard to excluded that we haven't missed anything. Any idea how to address the failing `make -j` problem?

Similar to how the dependency list is critical to the forward compilation of codes that `USE` modules, there needs to be an `*_tl.f90` or `*_ad.f90` dependency list generated.

jm-c commented:
@IvanaEscobar I've just updated issue #855 to reflect your latest commit.

mjlosch commented:
@IvanaEscobar found that In addressing the remaining problem of `make -j N` not working, the dependency list at the end of the Makefile has to augmented after the AD-code has been generated. There are specific requirements, e.g. the modules can only be used from other AD-generated code, so that calling `tools/f90mkdepend` again will give the wrong dependencies.

The first step is to write a specific script modelled after `tools/f90mkdepend` to give the correct dependencies. In a second step it may be worth it to add a "magic string" method (as in e.g. `xmakedepend`), so that a second call of this script will replace output from the previous call instead of appending it again. There's could be also something like the `-a` option in `xmakedepend`.

IvanaEscobar commented:
I added a simple script similar to `f90mkdepend`, called `f90mkdepend_ad`. This should handle `*_tl.f90/f` and `*_ad.f90/f` file dependencies by including a new file at the bottom of `Makefile`:
```
# DO NOT DELETE
...
include Makefile_ad.dep
```

The simple bash script **does not** consider module `USE`s of files such as `ad_taf_output.F` which are an MITgcm artifact for the user who wants to run TAF code all contained within a single file. @mjlosch had an idea on how to handle this that could be worth updating in this PR

IvanaEscobar commented:
There's a pesky file `ctrl_bound_tl.F` that was blocking `tools/f90mkdepend_taf` from working in an AD regime. Found a workaround. 

We now also account for `ftl_taf_outp*` and `ad_taf_outp*` in the generation of TAF file dependencies.

mjlosch reviewed (commented):
The current version leads to a lot of redundant dependencies. Makefile_taf.dep contains 549 word in 21 lines. I will push a version where the file has 100 words in 19 lines.

IvanaEscobar commented:
Updates to `f90mkdepend_taf` work for ihop set-up too.

can robustly apply parallel make targets associated with F90 TAF generated codes!

mjlosch commented:
If you would like to use the latest version, I can add something like this
```
# Determine if this is the first time, the Makefile is appened
makefile=Makefile
magic_string='# f90mkdepend_taf output starts here'
echo "" >> $makefile

# delete all lines from $magic_string to the end of the file
sed "/^$magic_string/,$ d" Makefile > Makefile.header
mv -f Makefile.header $makefile

# add the magic string
cat >> $makefile << END_OF_APPEND
$magic_string

END_OF_APPEND
# append makefile
cat $dep_file >> $makefile
```
to the end of `f90mkdepend_taf`, which makes it possible to directly append the Makefile (and do it multiple times without repeating everything)

IvanaEscobar commented:
> If you would like to use the latest version, I can add something like this
> 
> ```
> # Determine if this is the first time, the Makefile is appened
> makefile=Makefile
> magic_string='# f90mkdepend_taf output starts here'
> echo "" >> $makefile
> 
> # delete all lines from $magic_string to the end of the file
> sed "/^$magic_string/,$ d" Makefile > Makefile.header
> mv -f Makefile.header $makefile
> 
> # add the magic string
> cat >> $makefile << END_OF_APPEND
> $magic_string
> 
> END_OF_APPEND
> # append makefile
> cat $dep_file >> $makefile
> ```
> 
> to the end of `f90mkdepend_taf`, which makes it possible to directly append the Makefile (and do it multiple times without repeating everything)

I noticed that multiple instances of the `include` statement doesn't hurt the Makefile. If you prefer enforcing a single instance, then a magic string search and delete works perfectly!

Is there any other open issues with the `f90mkdepend_taf` inclusion that you are aware of?

mjlosch commented:
@IvanaEscobar You are right, using the "magic string" method is not really urgent here. I just wanted to document this idea. Maybe @jm-c will have an opinion.

Now `f90mkdepend_taf` is only used for the case of passing multiple files to TAF, the other two cases are not covered. The first case (just passing `ad_input_code.f`) does not need any post-processing because all code is in the resulting file. The case of passing `ad_input_code.f` and `ad_inpF90_code.f` is not entirely clear. I tried adding the current version of `f90mkdepend_taf` and it doesn't hurt, but it also works without it. Maybe we shouldn't spend time on a method that should be deprecated once more F90 and free format code is used in the MITgcm?

Currently I don't see any other issues with `f90mkdepend_taf`. I only tried it on one additional experiment so far, but I don't think that other experiments should cause any issues (but I didn't try the TL part).

IvanaEscobar commented:
Using the latest updates, the TL model compiles as well in my set-ups.

IvanaEscobar commented:
Merging the latest updates from the `master` branch still works. 

I was thinking `genmake2` may be faster for most builds if there was a check of whether or not F90 code even exists in the code. I'm unsure if it matters so much. Do y'all think it best to introduce a conditional for whether or not the user's build contains F90 code in this PR?

mjlosch commented:
> I was thinking genmake2 may be faster for most builds if there was a check of whether or not F90 code even exists in the code. I'm unsure if it matters so much. Do y'all think it best to introduce a conditional for whether or not the user's build contains F90 code in this PR?

Is that important? The most expensive part is still figuring out the dependencies for the `*.F` files, so I would not overload this PR and maybe have a subsequent one where you address speedup and beautification.

IvanaEscobar commented:
Sounds good

mjlosch commented:
I don't have anything new to add to this since my last review. I think it would be a good idea to have someone else have a look, e.g. @jm-c to see whether the general strategy is OK (i.e. have the extra file that is included per "include" in the makefile).

mjlosch reviewed (commented):
Still looks good to me. Small remaining questions (like multiple inclusions of the dependencies of TAF-generated files) can be done later, if desired.

IvanaEscobar commented:
Added an option to clean `Makefile` through `genmake2` target `Clean`

jm-c commented:
@IvanaEscobar I started to look at the changes in `genmake2`, but it's taking time. 
For now, I would suggest to update this PR description as several points are no longer relevant (e.g., in the list of points from issue #855, the second and third point listed here are not addressed in this PR ; and this comment also: "Any f90 codes in *.list needs to be updated from *.f90 to *.F90."  does not apply here).
And I saw you ihop git repos, but if I want to give it a try (to see how these changes work with modules in src code send to TAF), is there a small test that I can try ? Martin already asked about this (sorry for asking to repeat) but I did not find it in this PR discussion.

jm-c commented:
I am going to update issue #855 as this PR does not address task 6 there (third point in this PR description short list) but does address task 7 there (fourth point in this PR description short list) . Note that I am not suggesting to change this PR but, for now,  just to update the information.

jm-c commented:
I pushed an intermediate version of `genmake2`, called "genmake2_tmp", just to facilitate the comparison with both the original, master branch version and this PR new version, As mentioned in the commit, there is no intention at all to keep this file, it will be removed before merging.
Compared to original, master branch version, it contains the following:
 a) switch order the order (F90 first then F77 then C) in Makefile dependence and rules  (like in this PR), addressing task 7 of issue #885
b) remove the 2 "unused target" + some old commented out stuff (as in this PR)
 c) minor suffix change for FF90 processing (as in this PR).
From my point of view,  this is usefull as I can more clearly what the changes for handling modules with TAF as I can compare directly `genmake2` and the `Makefile` from this PR with this intermediate version generated one.

IvanaEscobar commented:
> @IvanaEscobar I started to look at the changes in `genmake2`, but it's taking time. For now, I would suggest to update this PR description as several points are no longer relevant (e.g., in the list of points from issue #855, the second and third point listed here are not addressed in this PR ; and this comment also: "Any f90 codes in *.list needs to be updated from *.f90 to *.F90." does not apply here). And I saw you ihop git repos, but if I want to give it a try (to see how these changes work with modules in src code send to TAF), is there a small test that I can try ? Martin already asked about this (sorry for asking to repeat) but I did not find it in this PR discussion.

I updated the PR description as you suggested. This PR does provide a way for F90 source code dependencies to be appended in the Makefile, so does that address the second point of this PR description @jm-c ? Otherwise, Ill update the description again the reflect the change this PR handles.

IvanaEscobar commented:
> And I saw you ihop git repos, but if I want to give it a try (to see how these changes work with modules in src code send to TAF), is there a small test that I can try ? Martin already asked about this (sorry for asking to repeat) but I did not find it in this PR discussion.

The most comprehensive test will be the `ihop+baroclinic_gyre` example. I believe Martin was using the [GMAO set-up](https://github.com/geos-mitgcm-ecco/gmao_mitgcm_coupling), which won't include modules.

jm-c commented:
@IvanaEscobar this second point (in this PR description, quoted from issue 855) is now marked as "done" in issue 855 task list, as a consequence of PR #865 beeing merged (sometime ago). This point was not specific to TAF, and not related to the use of modules, so I think it's fair to have it marked as done.

IvanaEscobar commented:
> @IvanaEscobar this second point (in this PR description, quoted from issue 855) is now marked as "done" in issue 855 task list, as a consequence of PR #865 beeing merged (sometime ago). This point was not specific to TAF, and not related to the use of modules, so I think it's fair to have it marked as done.

Sounds good, I've updated the description to highlight that my changes have to do with TAF code dependencies

jm-c commented:
Update: the ihop attempt I reported before was with MITgcm code from this branch (and not with checkpoint68s), so may be this is the issue ?

mjlosch commented:
@jm-c I think you need to use `-ncad` for the ihop-example with `tutorial_baroclinic_gyre` to work

IvanaEscobar commented:
As @mjlosch mentioned, using `-ncad` in your `genmake2` call should get you past these errors. I haven't run without this flag yet!

The code should compile without reverting the checkpoint68s, since the contents of `mitgcm_code` are up to date with the latest version of `MITgcm`. 

---
I have a forked branch that I use to compile with ihop here: https://github.com/IvanaEscobar/MITgcm/tree/ihop

It contains soft links to the repo `ihop/mitgcm_code` source codes and has been kept up with the various feature branches that remain un-merged such as this PR and #887 . This branch could help get past the TAF compile time error you're seeing, @jm-c

jm-c commented:
@IvanaEscobar I did not try yet your MITgcm ihop branch, but after removing `AUTODIFF_MYFIELDS.h` from `mitgcm_code/forward_step.F` and `mitgcm_code/the_main_loop.F` and using "-ncad" I can pass the "adtaf" step, still not compiling `ssp_mod_ad.f90` but the error there comes from TAF generated code, in "subroutine gcmsspmd", missing some declaration, so I guess it's fine for now ?

IvanaEscobar commented:
@jm-c Yes, that's a recomputation management issue that TAF is aware of. For now, I manually remove any tape generation inside TAF modified subroutines i.e. any subroutines ending in `md`.

The ihop AD code needs to be adjusted to apply recomputation in a better way. @mjlosch has helped in this effort here: https://github.com/IvanaEscobar/ihop/pull/26

IvanaEscobar commented:
> **NOTE:** Only works with `genmake2` flags `-nocat4ad | -dog4ad | -ncad | -dad`, which are for sending multiple files to TAF instead of a single concatenated file such as `ad_input_code.F`

Updates PR description to highlight TAF command limitation.

jm-c commented:
A quick summary of what we dicsussed today regarding next steps:

1.  @jm-c  will do a little cleaning
2. @IvanaEscobar  will update the PR description
3.  @jm-c  will report how this works or not with `pkg/atm_phys`
4. @jm-c will work on moving the pieces under:
> \# Sort the f90 files in topological order

into a new function within genmake2 (before the "Sequential part of script starts here") as it's used 2 times (-> could be cleaner and easier to maintain).

@IvanaEscobar If I missed something, please correct/edit this list.

jm-c commented:
Just few comments regarding last 2 commits:

1. commit f5a1bae861a55c41b33e6944debffb98d477eb1d was just a little bit of cleaning
2. the last commit fixed  errors I was getting when running `genmake2`, when a) no F90 files present or b) some F90 files but none using module, like:
> ../../../tools/genmake2: line 3150: /tmp/genmk_jmc_930255.F90Files: No such file or directory
> tsort: /tmp/genmk_jmc_930255.F90Graph: No such file or directory
> ../../../tools/genmake2: line 3159: /tmp/genmk_jmc_930255.F90noGraph: No such file or directory

    also a typo got fixed as one TMP files was not cleaned up.
3. and also in last commit, when sending multiple files to TAF (i.e., using "-ncad"), files: ad_taf_outpF90.$FS90 or ftl_taf_outpF90.$FS90 where not used, no new dependence compared to PR #709, therefore can safely remove this addition.

jm-c commented:
I am going to revert the last changed in `f90mkdepend_taf` (commit: 408fe5297b2e7b297b3b2b652fb4d96fe4517c57 , from Nov 19, "rm include .dep file in Makefile") and simplify/move when "include $dep_file" is added to the Makefile:

 - if I run "make adall" a second time, after making a small change to one of the file sent to TAF, it results in multiple "include Makefile_taf.dep" at the bottom of Makefile.
   It's not very clean, not very efficient neither, and the "f90mkdepend_taf -clean" only remove the last instance of it and not both ; after that, the Makefile does not work anymore as  `Makefile_taf.dep` was removed but one "include Makefile_taf.dep" remains:
> Makefile:8802: Makefile_taf.dep: No such file or directory
> make: *** No rule to make target 'Makefile_taf.dep'.  Stop.

 - the "-clean" option for `f90mkdepend_taf` only works when the makefile is `Makefile` (hard coded there in f90mkdepend_taf).
 - I prefer to keep the Makefile unchanged after "make depend", even if `Makefile_taf.dep` is updated later (and more frequently) after the TAF command.

To address these points, will add "include Makefile_taf.dep" only once, as part of "make depend"

jm-c commented:
I looked more carefully to this PR changes to `genmake2`, specially versus `genmake2_tmp` which
already contains the "more logical and consistent ordering inside targets: ".F77", ".F90", ".c", ".h" :

1. The list of F90 files to send to TAF ("AD_F90FILES" from tmp-file: $TMP.adF90Files) is re-ordered to  account for the use of modules ; this is necessary with current version of TAF, and works well with `pkg/ihop`.
2. After returning adjoint-routine (or TL routine) source code, the dependencies caused by using modules in these TAF generated files are added with `f90mkdepend_taf` into  `Makefile_taf.dep`, which is included in the `Makefile`. This is also necessary and does work well with `pkg/ihop`.
3. The full list of F90 files ("F90_SRC_FILES", from tmp-file: $TMP.F90srclist) is also re-ordered to account         for the use of modules. This is added to `genmake2` here: https://github.com/IvanaEscobar/MITgcm/blob/9e530b6aadd18a7dbe76826592331e141bb71558/tools/genmake2#L3133C1-L3172C37
    using tmp-file "$TMP.F90Files". But this "F90_SRC_FILES" list is only used to compile the forward executable "mitgcmuv" and, for that, the dependencies from the use of modules is handled by `f90mkdepend`. 
   So it's not clear why this has been added here. Was it something to work around `f90mkdepend` problem on MacOS (fixed in PR #858) ?
  
Anyway, I tried to comment out this line https://github.com/IvanaEscobar/MITgcm/blob/9e530b6aadd18a7dbe76826592331e141bb71558/tools/genmake2#L3168  to preserve the original list of files in $TMP.F90srclist and the `ihop` testing set-up works as well as with current genmake2 re-ordering, can generate the adjoint executable the same way as before.
So my impression is that we could remove this addition (part 3 above) from this PR modified genmake2.

IvanaEscobar commented:
> I am going to revert the last changed in `f90mkdepend_taf` (commit: [408fe52](https://github.com/MITgcm/MITgcm/commit/408fe5297b2e7b297b3b2b652fb4d96fe4517c57) , from Nov 19, "rm include .dep file in Makefile") and simplify/move when "include $dep_file" is added to the Makefile:
> 
> * if I run "make adall" a second time, after making a small change to one of the file sent to TAF, it results in multiple "include Makefile_taf.dep" at the bottom of Makefile.
>   It's not very clean, not very efficient neither, and the "f90mkdepend_taf -clean" only remove the last instance of it and not both ; after that, the Makefile does not work anymore as  `Makefile_taf.dep` was removed but one "include Makefile_taf.dep" remains:
> 
> > Makefile:8802: Makefile_taf.dep: No such file or directory
> > make: *** No rule to make target 'Makefile_taf.dep'.  Stop.
> 
> * the "-clean" option for `f90mkdepend_taf` only works when the makefile is `Makefile` (hard coded there in f90mkdepend_taf).
> * I prefer to keep the Makefile unchanged after "make depend", even if `Makefile_taf.dep` is updated later (and more frequently) after the TAF command.
> 
> To address these points, will add "include Makefile_taf.dep" only once, as part of "make depend"

The changes in [ee3ac9c](https://github.com/MITgcm/MITgcm/pull/862/commits/ee3ac9c1427ef16b7a51b89a9a5b02a5ebb87ee8) are what we were hoping to do in the long-term! Avoiding hard coding `Makefile` and `Makefile_taf.dep` in `f90mkdepend_taf` by moving arguments  into `genmake2` never occurred to me!

IvanaEscobar commented:
> 3. The full list of F90 files ("F90_SRC_FILES", from tmp-file: $TMP.F90srclist) is also re-ordered to account         for the use of modules. This is added to `genmake2` here: https://github.com/IvanaEscobar/MITgcm/blob/9e530b6aadd18a7dbe76826592331e141bb71558/tools/genmake2#L3133C1-L3172C37
>    using tmp-file "$TMP.F90Files". But this "F90_SRC_FILES" list is only used to compile the forward executable "mitgcmuv" and, for that, the dependencies from the use of modules is handled by `f90mkdepend`.
>    So it's not clear why this has been added here. Was it something to work around `f90mkdepend` problem on MacOS (fixed in PR [Generic file endings for f90mkdepend #858](https://github.com/MITgcm/MITgcm/pull/858)) ?
> 
> Anyway, I tried to comment out this line https://github.com/IvanaEscobar/MITgcm/blob/9e530b6aadd18a7dbe76826592331e141bb71558/tools/genmake2#L3168 to preserve the original list of files in $TMP.F90srclist and the `ihop` testing set-up works as well as with current genmake2 re-ordering, can generate the adjoint executable the same way as before. So my impression is that we could remove this addition (part 3 above) from this PR modified genmake2.

It wasn't due to MacOS file case insensitivities. 

The set-up of `genmake2` will work fine in the `ihop` setting. This is because `pkg/ihop` only has F90 source codes that `CONTAIN` subroutines inside of modules, i.e. no F90 source code that is module-free. There are two _hybrid_ use cases not covered by `pkg/ihop`:

1. F90 source code with no `MODULE`s
2. F90 source codes with AND without `MODULE`s

From what I can remember, the topologically ordered list, `F90_SRC_FILES`, covers these two cases for generation of `mitgcmuv` (forward executable).

jm-c commented:
Regarding the re-ordering of src files in  "F90_SRC_FILES":
1. It does not seems to be necessary (and again, this is used to build the forward executable; 
   for the adjoint/TLM, the list is split between src files sent to TAF and the rest of src code):
 - we currently have a good mix of modules used in fixed-format src code (*.F),
   with PTRACERS_ALLOW_DYN_STATE in `pkg/ptracers` (all src files that contains `PTRACERS_MOD.h`
   use the 2 modules `ptracers_dyn_state_mod` and `ptracers_dyn_state_data_mod`, and 
   S/R PTRACERS_INIT_FIXED_DYNAMIC inside module `ptracers_dyn_state_mod`) and also few
   modules used by/in fixed-format src code (*.F) in `pkg/atm_phys`.
 - for free-format src code, `pkg/atm_phys` is similar to pkg/ihop in the sense that
   all free-format subroutines are defined within a module. Just to check, I took subroutine
   LSCALE_COND_INIT out of `module lscale_cond_mod.F90` into it's own src file "lscale_cond_init.F90",
   made minor adjustments to the remaining `lscale_cond_mod.F90`, and was able to get the 
   correct dependencies in `Makefile` to build the forward executable.
 - we also have one free-format src code `cost_test_local.F90` (in `verification/hs94.1x64x5/code_ad`)
   that does not use any module.
   So, it seems to me that all cases are covered. @jahn do you agree or did I missed something ?
2. Because "F90_SRC_FILES" list is not specific to TAF adjoint/TLM processing, I suggest to
   leave this list unchanged in this PR (meaning, removing the re-ordering). If it turns out
   that this re-ordering is needed for some special cases that I missed, will bring it back in 
   a different PR. This way this PR is kept focused on Adjoint built (easier to document and fix
   if something goes wrong). @IvanaEscobar do you agree ?

IvanaEscobar commented:
> 2. Because "F90_SRC_FILES" list is not specific to TAF adjoint/TLM processing, I suggest to
>    leave this list unchanged in this PR (meaning, removing the re-ordering). If it turns out
>    that this re-ordering is needed for some special cases that I missed, will bring it back in
>    a different PR. This way this PR is kept focused on Adjoint built (easier to document and fix
>    if something goes wrong). @IvanaEscobar do you agree ?

Yes, let's remove the forward compilation re-ordering for now. If the reason becomes clear to me again, I'll report it to y'all.

jm-c commented:
OK, I will take care of this later today.

jm-c commented:
Regarding last commit [1f86763](https://github.com/MITgcm/MITgcm/pull/862/commits/1f867635bbe2d0d68a67cf2df42d207a950140c5), I just move the re-ordering into a new function, without making any change. I think it might be possible to improve a little bit this part, but it would also be simpler if we decide to change the content of the "\*_ad_diff.list" (to list "\*.F" and "\*.F90" files). I added a new task (number 13) in issue #855 related to this suffix change, but this would be for a future PR if we decide to do it.

jm-c commented:
I would like to remove the option to set the 2 new TAF related variables `TAF_F77_FLAGS` and `TAF_F90_FLAGS` from `genmake2` command line, i.e., "-taf_f77_flags" and "-taf_f90_flags":
At this point, it's not clear if this is very useful as these 2 new variables can be set either from a customized ad_optfile or in `genmake_local`. And adding new command line option to genmake without updating the top "Usage:" part (printed with "-help") seems a little bit half way.

And unrelated: since targets "adj_depend" and "ftl_depend" have no dependence, we could think of removing these 2 new targets and directly adding the line-command in place of "make  -f Makefile  adj_depend" and "make -f Makefile ftl_depend". This would be a little bit faster/simpler.

jm-c reviewed (approved):
LGTM

jm-c commented:
Regarding the attempt to generate Adjoint of exp. `atm_gray` (which uses `pkg/atm_phys`):
I don't have the intention to use or test `pkg/atm_phys` adjoint, but just to see if the modified `genmake2` here was working for this type of problem (with many modules in *.F90 source files).

I can build (and run) the forward executable with Makefile generated with "-ncad" but when trying "make adtaf", TAF returns an error related to the use of module inside a fixed format file, in my case `atm_phys_driver.f` which contains:
>       use surface_flux_mod
and a CALL to SURFACE_FLUX. The error is:

> *ERROR*   atm_phys_driver.f:1172 : cannot find specific function to surface_flux
> *ERROR*   atm_phys_driver.f:1172 : cannot find specific function to surface_flux
>
>  found 2 syntax error(s) in file atm_phys_driver.f

This does not affect this PR but just indicates that this pkg cannot be used to check how this PR works.

jm-c commented:
From my side, this PR is ready to be merged. @IvanaEscobar and @mjlosch do you agree ?
And just before merging I will remove my `genmake2_tmp` addition (and update tag-index).

IvanaEscobar commented:
@jm-c 

I just tried it out, and it seems to compile for me too. LGTM for merging

**NOTE:** There are TAF related errors addressed as an [ihop issue](https://github.com/IvanaEscobar/ihop/issues/33#issuecomment-2675073476) pending resolution. This only affects the `pkg/ihop` test case.

IvanaEscobar commented:
The TAF issue has been resolved by TAF dev, see [ihop issue comment](https://github.com/IvanaEscobar/ihop/issues/33#issuecomment-2678962526)

jm-c commented:
Will merge this PR tomorrow, unless someone wants to add something.

mjlosch commented:
Sorry, I don't have time to look at this thoroughly, but I think everything looks good. I assume that the `genmake2_tmp` (looks like a copy of the original `genmake2`) is still there on purpose; I just don't remember the discussion.
